### 1. Import the necessary library

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

### 2. Load the CSV data into DataFrame

In [2]:
#importing the dataset
df=pd.read_csv(r'C:\Users\USER\Desktop\fifa21 raw data v2.csv', low_memory=False)
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage  \
0           Left   93            RW   1-Jul-04           NaN  €103.5M  €560K   
1          Right   92            ST  10-Jul-18           NaN     €63M  €220K   
2          Right   91            GK  16-Jul-14           NaN    €120M  €125K   
3          Right   91           CAM  30-Aug-15           NaN    €129M  €370K   
4          Right   91            LW   3-Aug-17           NaN    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   
2        €159.4M         95        13         11                15   
3          €161M        407        94         82                55   
4        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalt

### 3.1 Display the column names

In [3]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

### 3.2 Inspect the shape of the dataset, in order to know the number of rows and columns.

In [4]:
df.shape

(18979, 77)

Observation: We've 18979 rows and 77 columns

### 3.3 Display the datatype of the columns using the .info()

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

### 3.4 Check for duplicate values in the dataset using .duplicated() 

In [6]:
df.duplicated().sum()

0

observation: There're no duplicate values in the dataset

### 3.5 check for missing values using .isnull() and .sum() to know the number of missing values in each column

In [7]:
df.isnull().sum()

ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64

Observation: The 'Loan Date End' column have  17966 and the 'Hits' column have 2595 missing values, while other columns do not have missing values

### 4. Drop the 'Loan Date Column' due to many NaN values

In [8]:
# Dropping the loan data column due to many NaN values
df.drop('Loan Date End',axis=1, inplace=True)
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position     Joined    Value   Wage Release Clause  \
0           Left   93            RW   1-Jul-04  €103.5M  €560K        €138.4M   
1          Right   92            ST  10-Jul-18     €63M  €220K         €75.9M   
2          Right   91            GK  16-Jul-14    €120M  €125K        €159.4M   
3          Right   91           CAM  30-Aug-15    €129M  €370K          €161M   
4          Right   91            LW   3-Aug-17    €132M  €270K        €166.5M   

   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  \
0        429        85         95                70             91       88   
1        437        84         95                90             82       86   
2         95        13         11                15             43       13   
3        407        94         82                55             94       82   
4        408        85         87                62             87       87   

   Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0    470         96     93           94            91            96       451   
1    414         88     81           76            77            92       431   
2    109         12     13           14            40            30       307   
3    441         88     85           83            93            92       398   
4    448         95     88           89            81            95       453   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   

   Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0       68       72        69          94        347          44   
1       95       84        78          93        353          63   
2       78       41        78          12        140          34   
3       63       89        74          91        408          76   
4       62       81        50          84        356          51   

   Interceptio

observation: The 'Loan Date End' column have been dropped

### 4.1 Strip '\n\n\n\n' from the club column and replace it with a string

In [9]:
# Stripping '\n\n\n\n' from the club column 
df['Club']=df['Club'].str.replace('\n\n\n\n',' ')
df['Club']

0                FC Barcelona
1                    Juventus
2             Atlético Madrid
3             Manchester City
4         Paris Saint-Germain
                 ...         
18974              Wuhan Zall
18975         Oldham Athletic
18976              Derry City
18977        Dalian YiFang FC
18978        Dalian YiFang FC
Name: Club, Length: 18979, dtype: object

### 5. Write a function to convert height of players in inches to cm on the 'Height'  column and apply the function to the column using .map()

In [10]:
#Convert inches to cm on the height column
def func(x):
    """ to convert inches to cm"""
    if '"' in x:
        y= x.replace('"', '') # to remove the symbol '"' and replace with an empty string
        y= y.split("\'") # to split it into feet and inches
        y=int(y[0])+int(y[1])*0.0833 
        y=30.48 * y   
        return int(round(y,0))
    else:
        return int(x.replace('cm',''))
df['Height']=df.Height.map(func)

### 5.1 Display the height of the first five players on the converted 'Height' column and use .describe() to describe the information on the height column

In [11]:
#print the converted 'height' column and describing it using the .describe method()
print(df.Height.head())
print('\n')
print(df.Height.describe())

0    170
1    187
2    188
3    181
4    175
Name: Height, dtype: int64


count    18979.000000
mean       181.200221
std          6.840054
min        155.000000
25%        176.000000
50%        181.000000
75%        186.000000
max        206.000000
Name: Height, dtype: float64


Observation: Now we've successfully converted the height of those players in inches to cm

### 6 Write a function to convert the weight of some players that's in kg to lbs on the 'Weight' column, and use .map() to apply the function on the weight column

In [12]:
#converting kg to lbs on the weight column
def kg_to_lbs(x):
    if 'kg' in x:
        y= x.replace('kg', '')
        y=int(y[0])*2.205 
        return int(round(y,0))
    else:
        return int(x.replace('lbs',''))
df['Weight']=df.Weight.map(kg_to_lbs)

### 6.1 Display the weight of the first five players on the converted 'Weight' column using the .head() method, and .describe() to describe the information on the weight column

In [13]:
#printing the converted values and describing the values
print(df.Weight.head())
print(df.Weight.describe())

0    15
1    18
2    18
3    15
4    13
Name: Weight, dtype: int64
count    18979.000000
mean        15.717372
std          7.405818
min          2.000000
25%         15.000000
50%         15.000000
75%         18.000000
max        203.000000
Name: Weight, dtype: float64


Observation: Now we've successfully converted the weight of those players in kg to lbs

### 7 Strip '€' sign from the 'Value' column using .replace(). Write a function to convert those values in 'K' to thousands and those in 'M' to millions. Then use .apply() to effect the function on the 'Value' column

In [14]:
#standardizing the value column
df['Value']=df['Value'].str.replace('€', '')
def conversion(value):
    if value.endswith('K'):
        return float(value[:-1])*1000
    elif value.endswith('M'):
        return float(value[:-1])*10000
    else:
        return float(value)
df['Value']=df['Value'].apply(conversion)

### 7.1 Standardize the 'Value' column by divding the column by 1000000. Rename the column to indicate its in million and print the standardized column

In [15]:
df['Value_M']=df['Value']/1000000
print(df['Value_M'])

0        1.035
1        0.630
2        1.200
3        1.290
4        1.320
         ...  
18974    0.100
18975    0.130
18976    0.120
18977    0.100
18978    0.100
Name: Value_M, Length: 18979, dtype: float64


### 8. Strip '€' sign and 'K' sign from the 'Wage' column, using .replace(). Rename the column to indicate that the players wage are in thousand

In [16]:
df['Wage_K']=df['Wage'].str.replace('€','').str.replace('K','')
df['Wage_K']

0        560
1        220
2        125
3        370
4        270
        ... 
18974      1
18975    500
18976    500
18977      2
18978      1
Name: Wage_K, Length: 18979, dtype: object

### 8.1 Convert the 'Wage_K' column from object to int using the .astype()

In [17]:
df['Wage_K'].astype(int)

0        560
1        220
2        125
3        370
4        270
        ... 
18974      1
18975    500
18976    500
18977      2
18978      1
Name: Wage_K, Length: 18979, dtype: int32

In [18]:
# Drop 'Value' and 'Wage' columns leaving only the rename ones
df.drop(['Value', 'Wage'], axis=1, inplace=True)
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Release Clause',
       'Attacking', 'Crossing', 'Finishing', 'Heading Accuracy',
       'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve',
       'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 'PHY', 'Hits', 'Value_M', 'Wage_K'

### 8.2 Strip '€' sign from the 'Release Clause' column using .replace(). Write a function to convert those values in 'K' to thousands and those in 'M' to millions. Then use .apply() to effect the function on the 'Release Clause' column

In [19]:
df['Release Clause']=df['Release Clause'].str.replace('€', '')
def conversion1(Release_Clause):
    if Release_Clause.endswith('K'):
        return float(Release_Clause[:-1])*1000
    elif Release_Clause.endswith('M'):
        return float(Release_Clause[:-1])*10000
    else:
        return float(Release_Clause)
df['Release Clause']=df['Release Clause'].apply(conversion1)

### 8.3 Standardize the 'Release Clause' column by divding the column by 1000000. Rename the column to 'Release Clause_M' to indicate its in million and print the standardized column

In [20]:
#Converting clause column from object to float
df['Release Clause_M']=df['Release Clause']/1000000
df['Release Clause_M']

0        1.384
1        0.759
2        1.594
3        1.610
4        1.665
         ...  
18974    0.070
18975    0.165
18976    0.131
18977    0.088
18978    0.079
Name: Release Clause_M, Length: 18979, dtype: float64

### 9 strip '★' from the 'W/F', 'SM' and 'IR' column which contains different aspect of players rating with a ranking from 1-5. Then convert columns to integer

In [21]:
#stripping '★' from the 'W/F' column
df['W/F']=df['W/F'].str.replace('★','')
df['W/F']

0        4 
1        4 
2        3 
3        5 
4        5 
         ..
18974    2 
18975    2 
18976    2 
18977    3 
18978    3 
Name: W/F, Length: 18979, dtype: object

In [22]:
#converting'W/F_★' from object to int  
df['W/F'].astype(int)

0        4
1        4
2        3
3        5
4        5
        ..
18974    2
18975    2
18976    2
18977    3
18978    3
Name: W/F, Length: 18979, dtype: int32

In [23]:
#Stripping '★' from SM column  
df['SM']=df['SM'].str.replace('★','')
df['SM']

0        4
1        5
2        1
3        4
4        5
        ..
18974    2
18975    2
18976    2
18977    2
18978    2
Name: SM, Length: 18979, dtype: object

In [24]:
#converting'SM_★' from object to int  
df['SM'].astype(int)

0        4
1        5
2        1
3        4
4        5
        ..
18974    2
18975    2
18976    2
18977    2
18978    2
Name: SM, Length: 18979, dtype: int32

In [25]:
#Stripping '★' from IR column  
df['IR']=df['IR'].str.replace('★','')
df['IR']

0        5 
1        5 
2        3 
3        4 
4        5 
         ..
18974    1 
18975    1 
18976    1 
18977    1 
18978    1 
Name: IR, Length: 18979, dtype: object

In [26]:
#converting'IR_★' from object to int  
df['IR'].astype(int)

0        5
1        5
2        3
3        4
4        5
        ..
18974    1
18975    1
18976    1
18977    1
18978    1
Name: IR, Length: 18979, dtype: int32

Observation: The star symbol have been removed from 'W/F', 'SM' and 'IR' columns and the columns converted to integer

### 9.1 Display the first five rows of the 'FIFA 21' dataset using .head()

In [27]:
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                  Club     Contract   Positions  Height  Weight  \
0   93          FC Barcelona  2004 ~ 2021  RW, ST, CF     170      15   
1   92              Juventus  2018 ~ 2022      ST, LW     187      18   
2   93       Atlético Madrid  2014 ~ 2023          GK     188      18   
3   91       Manchester City  2015 ~ 2023     CAM, CM     181      15   
4   91   Paris Saint-Germain  2017 ~ 2022     LW, CAM     175      13   

  Preferred Foot  BOV Best Position     Joined  Release Clause  Attacking  \
0           Left   93            RW   1-Jul-04       1384000.0        429   
1          Right   92            ST  10-Jul-18        759000.0        437   
2          Right   91            GK  16-Jul-14       1594000.0         95   
3          Right   91           CAM  30-Aug-15       1610000.0        407   
4          Right   91            LW   3-Aug-17       1665000.0        408   

   Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  Skill  \
0        85         95                70             91       88    470   
1        84         95                90             82       86    414   
2        13         11                15             43       13    109   
3        94         82                55             94       82    441   
4        85         87                62             87       87    448   

   Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0         96     93           94            91            96       451   
1         88     81           76            77            92       431   
2         12     13           14            40            30       307   
3         88     85           83            93            92       398   
4         95     88           89            81            95       453   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   

   Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0       68       72        69          94        347          44   
1       95       84        78          93        353          63   
2       78       41        78          12        140          34   
3       63       89        74          91        408          76   
4       62       81        50          84        356          51   

   Interceptions  Positioning  Vision  Penalties  Composure  Defending  \
0             40           93      95         75         96 

In [28]:
#Displaying the column
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Release Clause',
       'Attacking', 'Crossing', 'Finishing', 'Heading Accuracy',
       'Short Passing', 'Volleys', 'Skill', 'Dribbling', 'Curve',
       'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 'PHY', 'Hits', 'Value_M', 'Wage_K'

### 10. Convert the 'Joined' column from object to datetime because column indicates the date the players joined their current team in FIFA 21 

In [29]:
df['Joined']=pd.to_datetime(df['Joined'])
df['Joined']

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: Joined, Length: 18979, dtype: datetime64[ns]

### 11 Check for unique values in the 'Hits' column

In [30]:
df['Hits'].unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

Observation: Some of the values in the 'Hits' column have letter 'K' attached to them, so we have remove the letter

### 11.1 Strip K from the 'Hits' column and rename the column to indicates its in thousand

In [45]:
df['Hits_K']=df['Hits'].str.replace('K','')
df['Hits_K']

0        771
1        562
2        150
3        207
4        595
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: Hits_K, Length: 18979, dtype: object

### 11.2 Fill the NaN values in the 'Hits' column with zero

In [43]:
df['Hits_K']=df['Hits'].fillna(0)
df['Hits_K']

0        771
1        562
2        150
3        207
4        595
        ... 
18974      0
18975      0
18976      0
18977      0
18978      0
Name: Hits_K, Length: 18979, dtype: object

In [44]:
#Checking if there're still missing values in the 'Hits'
df['Hits_K'].isnull().sum()

0

There're no more missing values on the 'Hits' column

### 11.3 Change the 'Hits' column datatype from object to float

In [35]:
df['Hits_K'].astype(float)

0        771.0
1        562.0
2        150.0
3        207.0
4        595.0
         ...  
18974      0.0
18975      0.0
18976      0.0
18977      0.0
18978      0.0
Name: Hits, Length: 18979, dtype: float64

### 12. write a function to split the 'Contract' column which contains the contract details of the players into    'Contract_start_year' and 'Contract_end_year'

In [36]:
#splitting the contract column into contract_start_year and contract_end_year
def main(k):
    start=[]
    gen=k.to_list()
    for i in gen:
        start.append(i[:7])
    return start

df['Contract_Year_Start']=main(df['Contract'])

def main1(k):
    end=[]
    gen=k.to_list()
    for i in gen:
        end.append(i[-7:])
    return end
df['Contract_Year_End']=main1(df['Contract'])
df['Contract_Year_End']=df['Contract_Year_End']

### 12.1 print the contract_start_year column

In [37]:
df['Contract_Year_Start']=df['Contract_Year_Start'].str.replace('~','')
df['Contract_Year_Start']

0        2004  
1        2018  
2        2014  
3        2015  
4        2017  
          ...  
18974    2018  
18975    2020  
18976    2019  
18977    2020  
18978    2019  
Name: Contract_Year_Start, Length: 18979, dtype: object

### 12. 2 print the contract_end_year column

In [38]:
df['Contract_Year_End']=df['Contract_Year_End'].str.replace('~','')
df['Contract_Year_End']

0          2021
1          2022
2          2023
3          2023
4          2022
          ...  
18974      2022
18975      2021
18976      2020
18977      2022
18978      2023
Name: Contract_Year_End, Length: 18979, dtype: object

### Dropping the 'Contract' after splitting into 'Contract_Year_Start' and 'Contracr_Year_End'

In [39]:
df.drop('Contract',axis=1,inplace=True)

In [40]:
#print the dataset
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                  Club   Positions  Height  Weight Preferred Foot  BOV  \
0   93          FC Barcelona  RW, ST, CF     170      15           Left   93   
1   92              Juventus      ST, LW     187      18          Right   92   
2   93       Atlético Madrid          GK     188      18          Right   91   
3   91       Manchester City     CAM, CM     181      15          Right   91   
4   91   Paris Saint-Germain     LW, CAM     175      13          Right   91   

  Best Position     Joined  Release Clause  Attacking  Crossing  Finishing  \
0            RW 2004-07-01       1384000.0        429        85         95   
1            ST 2018-07-10        759000.0        437        84         95   
2            GK 2014-07-16       1594000.0         95        13         11   
3           CAM 2015-08-30       1610000.0        407        94         82   
4            LW 2017-08-03       1665000.0        408        85         87   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   
1                90             82       86    414         88     81   
2                15             43       13    109         12     13   
3                55             94       82    441         88     85   
4                62             87       87    448         95     88   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   
1           76            77            92       431            87   
2           14            40            30       307            43   
3           83            93            92       398            77   
4           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         9

###  Rename 'LongName', '↓OVA' and 'POT' column to have explanatory columns

In [41]:
df.rename(columns={'LongName':'Full_Name', '↓OVA':'Overall_Rating%','POT':'Potential_Rating%'}, inplace=True)

#print the dataset
df.head()

ID               Name                     Full_Name  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   

   Overall_Rating%  Potential_Rating%                  Club   Positions  \
0               93                 93          FC Barcelona  RW, ST, CF   
1               92                 92              Juventus      ST, LW   
2               91                 93       Atlético Madrid          GK   
3               91                 91       Manchester City     CAM, CM   
4               91                 91   Paris Saint-Germain     LW, CAM   

   Height  Weight Preferred Foot  BOV Best Position     Joined  \
0     170      15           Left   93            RW 2004-07-01   
1     187      18          Right   92            ST 2018-07-10   
2     188      18          Right   91            GK 2014-07-16   
3     181      15          Right   91           CAM 2015-08-30   
4     175      13          Right   91            LW 2017-08-03   

   Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0       1384000.0        429        85         95                70   
1        759000.0        437        84         95                90   
2       1594000.0         95        13         11                15   
3       1610000.0        407        94         82                55   
4       1665000.0        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           